---
title: "Accessing NCBI data with the Rentrez package"
author: "Irmarie Fraticelli-Rodriguez"
date: "2023-05-15"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
### What is Rentrez?

Rentrez is an R interface that allows its users to interact with NCBI API. 
With Rentrez, you do not need to use any additional program or terminal to access NCBI Data. 
This means that you can request data from multiple databases (PubMed, SNP,Clinvar,SRA, Gene and others) in the same RStudio Session. 

If you want to learn more about Rentrez, you can access this [Rentrez Tutorial](https://cran.r-project.org/web/packages/rentrez/vignettes/rentrez_tutorial.html)

### Install and Load Rentrez

Now, remember that you need to install any new package that you want to use in RStudio. 
Also, once you have the package you need to load it. 
How can we do that?

```{r}

#install.packages('rentrez')
library(rentrez)

```

Now that we know that we have installed the package and load it into Rstudio. 
Lets take a look at the documentation available about this this resource.

Lets visit the [Rentrez Documentation](https://cran.r-project.org/web/packages/rentrez/rentrez.pdf)

This pdf provides details about each function that the R Package Rentrez has. 
For each command you may find a description, usage example, arguments and return value type. 
Remember that you can always use the helper(<command>) in R to search command descriptions. 

## Rentrez Fuctions

### Helper Functions

The fuctions listed below helps you learn a little bit more about Rentrez and NCBI Databases. 

| Function      | Usage       |
| ------------- |-------------|
| entrez_dbs()            | List of Entrez Databases |
| entrez_db_summary()     | Brief description of what the database is |
| entrez_db_searchable()  | Set of search terms that can used with this database |
| entrez_db_links()       | Set of databases that might contain linked records |
| extract_from_esummary() | Extract elements from a list of esummary records |

### entrez_dbs()

Lets see what entrez_dbs() returns: 

```{r}

entrez_dbs()

```
The function entrez_dbs() returns a list of databases that we can query. 

### entrez_db_summary()

How do we know that the database is up to date? 
We can use the entrez_db_summary function to see the latest updates of the selected NCBI database. 

```{r}

#entrez_db_summary([YOUR ANSWER HERE])

```
### entrez_db_searchable()

How can I build PubMed queries? What search fields does the database has?
We can use the entrez_db_searchable function to see searchable fields and its description. 


```{r}

#entrez_db_searchable([YOUR ANSWER HERE])

```

Lets strategize our search!
Now that we know the search fields you can use to build a query, lets perform our first search. 

## Performing searches in NCBI API

### entrez_search()

Similar to PubMed, Rentrez allows you to perform simple or boolean searches using the same structure that you would use in the PubMed search bar. 
The allowed boolean terms are **AND, OR, and NOT**.

First lets learn the syntax for a simple and a boolean search:

Simple Search:

>entrez_search(db= [database name], term= "searchword[field]")

Boolean search:

>entrez_search(db= [database name], term= "searchword[field]" <boolean term> "searchword[field]")

The entrez_search function takes 5 arguments but you don't need to use them all, but the db and term arguments are required.  

| Argument      | Type         | Use           |
| ------------- |------------- |-------------  |
| db            | CHARACTER    | Name of the database to search for |
| term          | CHARACTER    | The search term. In here you can use MESH terms, or any of the search fields by using PubMed Query Syntax |
| retmode       | VECTOR       | To select the format of your output (XML or JSON), by default will be XML  |
| use_history   | LOGICAL      | To store a history of searches in NCBI's server |


EXAMPLES: 

```{r}

#([YOUR ANSWER HERE]) <- entrez_search([YOUR ANSWER HERE])
#([YOUR ANSWER HERE]) <- entrez_search([YOUR ANSWER HERE])
#([YOUR ANSWER HERE]) <- entrez_search([YOUR ANSWER HERE])

```

#### What the entrez_search() function returns 

The entrez_search fuction returns a list of 5 elements. 
What are those elements? What they mean?

| Value         | Type          | Description   |
| ------------- | ------------- | ------------- |
| ids           | INTEGER | is an identifying number for each publication record |
| count         | INTEGER | is the number of publication records retrieved |
| retmax        | INTEGER | controls the maximum number of returned values, if this is not set the default value is 20 |
| web_history   | WEB HISTORY OBJECT | A key of your stored history search |
| QueryTranslation | CHARACTER       | search term as the NCBI interpreted it |
| file          | XMLInternalDocument | type of file resulted from search |
  
An important field here is the QueryTranslation value which allows you to see how NCBI API has interpreted your selected term words according to MESH Terms.

### entrez_summary()

Now lets retrieve basic information about the records we collected in one of our searches. 
But, first lest learn the syntax of the entrez_summary function. 

>entrez_summary(db="", id=, rettype = "")

The entrez_summary function takes 7 arguments but you dont need to use them all. Those are: 

| Argument      | Type          | Use           |
| ------------- | ------------- | ------------- |
| db           | CHARACTER | Name of the database to search for |
| id           | VECTOR  | Unique ID(s) for records in database  |
| web_history  | WEB HISTORY OBJECT | A web_history object |
| version      | CHARACTER | Either 1.0 or 2.0 |
| always_return_list | LOGICAL | Return a list of esummary objects even when only one ID is provided|
| retmode      | CHARACTER | By default, xml will be used for version 1.0 records, json for version 2.0. |
| config       | VECTOR | Configuration options passed to httr::GET |

EXAMPLES:

```{r}

#([YOUR ANSWER HERE]) <- entrez_summary([YOUR ANSWER HERE])
#([YOUR ANSWER HERE]) <- entrez_summary([YOUR ANSWER HERE])
#([YOUR ANSWER HERE]) <- entrez_summary([YOUR ANSWER HERE])

```

#### What the entrez_summary() function returns 

This simple search returns a list of 5 elements. 
What are those elements? What they mean?

| Value         | Type          | Description   |
| ------------- | ------------- | ------------- |
| eSummary Records          | LIST | A list of esummary records (if multiple IDs are passed and always_return_list if FALSE) or a single record |
| file |XMLInternalDocument | xml file containing the entire record returned by the NCBI |


### entrez_fetch() 

Another way to get records from the NBCI API, is by using the entrez_fetch function. entrez_summary and entrez_fetch are almost the same function, both of them will help you extract data from the NCBI API. But the difference between entrez_summary and entrez_fetch is in the way both functions arranges the data in RStudio. The entrez_summary returns a dataframe with lists of records while entrez_fetch returns an XML document. 

For the entrez_summary function, rentrez has created a function that allows you to extract information from each metadata field avoiding the challenges that comes with navigating an XML parsed documents.  

EXAMPLE:
```{r}

#([YOUR ANSWER HERE]) <- entrez_fetch([YOUR ANSWER HERE])

```

Now that we now how to get data full records in Rentrez using entrez_esummary and entrez_fetch, lets learn a helper function that will allows us to get specific data fields from 
NCBI API XML document. 

### extract_from_esummary()

This function helps you navigate through an XMLInternalDocument and extract elements from a list of esummary records. 
But, first lest learn the syntax of the extract_from_esummary function. 

>extract_from_esummary(esummaries, elements, simplify = TRUE)

The extract_from_summary function takes 3 arguments but you dont need to use them all. Those are: 

| Argument      | Type          | Use           |
| ------------- | ------------- | ------------- |
| esumaries     | ENTREZ_SUMMARY ITEM | Either an esummary or an esummary_list (as returned by entrez_summary) |
| elements      | CHARACTER           | Unique ID(s) for records in database  |
| simplify      | LOGICAL             | TODO If possible return a vector |

EXAMPLE:

```{r}

# If you have been performing searches in PubMed, you can use the following coding lines as guide.

#uids <- extract_from_esummary([YOUR ANSWER HERE])
#authors <- extract_from_esummary([YOUR ANSWER HERE])
#pubdate <- extract_from_esummary([YOUR ANSWER HERE])
#title <- extract_from_esummary([YOUR ANSWER HERE])
#pubtype <- extract_from_esummary([YOUR ANSWER HERE])
#source <- extract_from_esummary([YOUR ANSWER HERE])
#issue <- extract_from_esummary([YOUR ANSWER HERE])
#volume <- extract_from_esummary([YOUR ANSWER HERE])
#pages <- extract_from_esummary([YOUR ANSWER HERE])

```

#### What the extract_from_esummary() function returns 

| Value         | Type          | Description   |
| ------------- | ------------- | ------------- |
| e_summary item | LIST OR VECTOR | A list or vector of the requested item. |

### entrez_global_query()

### entrez_links()

SEARCH RELATED ARTICLES USING ENTREZ_LINK
Searching for related articles in other database
Evaluating links for PCOS between NCBI databases
Lets select an article that is related with two characteristics of PCOS, Insulin Resistance and Metabolic Syndrome

Selected Article 
Genetics of Insulin Resistance and the Metabolic Syndrome, https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4911377/
PMCID: PMC4911377
PMID: 27312935

Our results shows that there are related articles in databases such as the Single Nucleotide Polymorphism Database(dbSNP), Medgen, the Taxonomy database and others
pcos_links <-entrez_link(dbfrom="pubmed", id="10.1007/s11886-016-0755-4", db= "all")

Searching the titles for articles that cited the selected article
summary_pcos_citedin <- entrez_summary(db="pubmed", id=pcos_links$links$pubmed_pubmed_citedin, rettype = "xml")
Searching for matches in the SRA Database
summary_pcos_sra <- entrez_summary(db="sra", id=pcos_links$links$pubmed_sra, rettype = "xml")
Searching for matches in the SNP Database
summary_pcos_snp <- entrez_summary(db="snp", id=pcos_links$links$pubmed_snp, rettype = "xml")

